tests/crud_operations.js

```diff
@@ -0,0 +1,133 @@
+    let res = await req;
+    
+    console.log("Response Text:", res.text);  // Log the response to check CSRF token presence
+    
+    const textNoLineEnd = res.text.replaceAll("\n", "");
+    
+    // Extract CSRF token from the form
+    this.csrfToken = /_csrf"\s+value="(.*?)"/.exec(textNoLineEnd);
+    
+    if (!this.csrfToken) {
+      throw new Error("CSRF Token not found in the response");
+    }
+    
+    // Check that the CSRF token is present and correct
+    console.log("Extracted CSRF Token:", this.csrfToken[1]);
+    this.csrfToken = this.csrfToken[1];  // Extracted CSRF token value
+    
+    let cookies = res.headers["set-cookie"];
+    this.csrfCookie = cookies.find((element) => element.startsWith("_csrf"));
+    this.sessionCookie = cookies.find((element) => element.startsWith("connect.sid"));
+    
+    if (!this.csrfCookie || !this.sessionCookie) {
+      throw new Error("Cookies are not set correctly.");
+    }
+    
+    // Prepare data to post (login)
+    const dataToPost = {
+      email: this.test_user.email,
+      password: testUserPassword,
+      _csrf: this.csrfToken,
+    };
+    
+    // Log CSRF token and session cookies before making the login POST request
+    console.log("Sending CSRF Token:", this.csrfToken);
+    console.log("Cookies:", this.csrfCookie, this.sessionCookie);
+    
+    // Send the login request
+    req = request
+      .execute(app)
+      .post("/sessions/logon")
+      .set("Cookie", `${this.csrfCookie}; ${this.sessionCookie}`)
+      .set("content-type", "application/x-www-form-urlencoded")
+      .redirects(1)
+      .send(dataToPost);
+    
+    res = await req;
+    console.log("Response after login:", res.text);  // Log the entire respons
+
+    // Log the cookies to check if the session cookie is set
+console.log("Cookies after login:", res.headers['set-cookie']);
+
+    // Check for redirect status code (should be 3xx for redirect)
+  console.log("Status Code after login:", res.status);  // Inspect the status code
+
+  // If redirected, we should check the CSRF token again in the next page
+  if (res.status >= 300 && res.status < 400) {
+    req = request.execute(app).get(res.headers.location).set("Cookie", res.headers["set-cookie"]).send();
+    res = await req;
+    const textNoLineEndAfterLogin = res.text.replaceAll("\n", "");
+    this.csrfToken = /_csrf"\s+value="(.*?)"/.exec(textNoLineEndAfterLogin);
+
+    if (!this.csrfToken) {
+      throw new Error("CSRF Token not found in the response after login redirect");
+    }
+
+    console.log("Extracted CSRF Token after login redirect:", this.csrfToken[1]);
+    this.csrfToken = this.csrfToken[1]; // Update the CSRF token post-login
+
+    // Check for session cookie after login
+    cookies = res.headers["set-cookie"];
+    this.sessionCookie = cookies.find((element) => element.startsWith("connect.sid"));
+  }
+
+    
+    expect(this.csrfToken).to.not.be.undefined;
```

Comment:
You'd instead want to move this up, and potentially check for more values like `null` if this does only check for `undefined`.